### PR TITLE
Fix the problem that `DateTime.difference` reports an error on Flutter Web.

### PR DESCRIPTION
**Pull request by ipcjs**
Fix the bug on Flutter Web:
image

### CODE REVIEW

File: lib/src/const_date_time.dart [hunk 1]
```diff
@@ -195,4 +195,19 @@ class ConstDateTime implements DateTime {
...
+  @override
+  dynamic noSuchMethod(Invocation invocation) {
+    // On the Web platform, the [DateTime.difference] method reads the `other._value` variable internally.
+    // ConstDateTime does not have a `_value` variable, which leads to errors when running
+    // Returning [millisecondsSinceEpoch] in this case will prevent the error from being reported.
+    //
+    // @see https://github.com/dart-lang/sdk/blob/56e581aa3415858bbaf22bee23be705a68f8a03e/sdk/lib/_internal/js_runtime/lib/core_patch.dart#L342
+    if (invocation.isGetter
+        // Flutter web performs minified in release mode, resulting in not getting the correct memberName.
+        /* && invocation.memberName == #_value */) {
```
Comment (westy92) [inline lib/src/const_date_time.dart:210]: Could you please try:
```dart
invocation.memberName == const Symbol('_value')) {
```
in release mode on web?

File: lib/src/const_date_time.dart [hunk 2]
```diff
@@ -195,4 +195,19 @@ class ConstDateTime implements DateTime {
...
+      return dateTime.millisecondsSinceEpoch;
```
Comment (westy92) [inline lib/src/const_date_time.dart:211]:
```dart
return dateTime.microsecondsSinceEpoch;
```

**Comment by westy92** [comment]
Thank you for your contribution! I left a few comments.

File: example_flutter/lib/main.dart [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    title: title,
+    home: Scaffold(
+      appBar: AppBar(
+        title: const Text(title),
+      ),
+      body: Center(
+        child: Column(
+          mainAxisSize: MainAxisSize.min,
+          children: [
+            Text(now.difference(DateTime(0)).toString()),
```
Comment (westy92) [inline example_flutter/lib/main.dart:17]: Just an FYI, a workaround might be swapping the parameters and negating the value:
```dart
Text(-(const ConstDateTime(0)).difference(now).toString()),
```

File: lib/src/const_date_time.dart (hunk 2 above)
Comment (ipcjs) [reply lib/src/const_date_time.dart:211]: On JavaScript it is equal to [millisecondsSinceEpoch].

@see https://github.com/dart-lang/sdk/blob/main/sdk/lib/core/date_time.dart#L167

File: lib/src/const_date_time.dart (hunk 1 above)
Comment (ipcjs) [reply lib/src/const_date_time.dart:210]: `#_value` is a Symbol literals, It is the same object as `const Symbol('_value')`.

File: example_flutter/lib/main.dart (hunk 3 above)
Comment (ipcjs) [reply example_flutter/lib/main.dart:17]: In my code, I mix `ConstDateTime` and `DateTime`, and there is no way to guarantee the order of the calls.

File: lib/src/const_date_time.dart (hunk 1 above)
Comment (westy92) [reply lib/src/const_date_time.dart:210]: Ah, good to know. Thank you!

File: pubspec.yaml [hunk 4]
```diff
@@ -19,5 +19,10 @@ platforms:
   windows:
 
 dev_dependencies:
+  flutter:
```
Comment (ipcjs) [inline pubspec.yaml]: `const_date_time` is a Dart package. It should not depend on Flutter.

Comment (westy92) [reply pubspec.yaml]: Right. I was just trying to resolve the analysis errors. The correct approach was to exclude the new example altogether.

File: lib/src/const_date_time.dart [hunk 5]
```diff
@@ -195,4 +195,22 @@ class ConstDateTime implements DateTime {
...
   }
+
+  @override
+  dynamic noSuchMethod(Invocation invocation) {
+    // On the Web platform, the [DateTime.difference] method reads the `other._value` variable internally.
+    // ConstDateTime does not have a `_value` variable, which leads to errors when running
+    // Returning [millisecondsSinceEpoch] in this case will prevent the error from being reported.
+    //
+    // @see https://github.com/dart-lang/sdk/blob/56e581aa3415858bbaf22bee23be705a68f8a03e/sdk/lib/_internal/js_runtime/lib/core_patch.dart#L342
+    if (_kIsWeb && invocation.isGetter
```
Comment (westy92) [inline lib/src/const_date_time.dart]: Should we strengthen this check by adding `&& !invocation.isSetter`? I believe final properties follow this pattern.

**Review by westy92 — APPROVED** [review]
Looks good to me! I left one final question for another potential improvement.

**Comment by ipcjs** [comment]
I added some test code, unfortunately this solution only passes the test in the stable channel, maybe we should commit a pr to [@dart-lang/sdk](https://github.com/dart-lang/sdk)?

截屏2022-12-15 17 28 40

**Comment by westy92** [comment]
Filed issue https://github.com/dart-lang/sdk/issues/50750.

**Comment by westy92** [comment]
I'm happy to land this patch even though it will deliberately break in Dart 2.19. Once Dart 2.19 is released, we can re-evaluate our available options.

**Comment by westy92** [comment]
Published in version 1.0.1. Thank you for your contribution!

**Pull request merged**